File: apps/mobile/android/app/src/main/java/com/enflame/happy/ui/viewmodel/PurchaseViewModel.kt
package com.enflame.happy.ui.viewmodel

import android.app.Activity
import android.util.Log
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.enflame.happy.data.billing.PurchaseService
import com.enflame.happy.domain.model.BillingPeriod
import com.enflame.happy.domain.model.PurchaseResult
import com.enflame.happy.domain.model.SubscriptionPlan
import com.enflame.happy.domain.model.SubscriptionStatus
import com.enflame.happy.domain.model.SubscriptionTier
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.combine
import kotlinx.coroutines.flow.stateIn
import kotlinx.coroutines.launch
import javax.inject.Inject

/**
 * UI state for the paywall and subscription screens.
 *
 * @property availablePlans All available subscription plans from Google Play.
 * @property filteredPlans Plans filtered by the selected billing period.
 * @property subscriptionStatus The current subscription status.
 * @property selectedBillingPeriod The currently selected billing period filter.
 * @property isPurchasing Whether a purchase is in progress.
 * @property isLoading Whether products are being loaded.
 * @property isRestoring Whether a restore is in progress.
 * @property showPurchaseSuccess Whether to show the purchase success dialog.
 * @property showRestoreSuccess Whether to show the restore success dialog.
 * @property errorMessage Error message to display, or null.
 */
data class PurchaseUiState(
    val availablePlans: List<SubscriptionPlan> = emptyList(),
    val filteredPlans: List<SubscriptionPlan> = emptyList(),
    val subscriptionStatus: SubscriptionStatus = SubscriptionStatus.Loading,
    val selectedBillingPeriod: BillingPeriod = BillingPeriod.MONTHLY,
    val isPurchasing: Boolean = false,
    val isLoading: Boolean = true,
    val isRestoring: Boolean = false,
    val showPurchaseSuccess: Boolean = false,
    val showRestoreSuccess: Boolean = false,
    val errorMessage: String? = null
) {
    /** Whether the user currently has an active subscription. */
    val isSubscribed: Boolean
        get() = subscriptionStatus.isActive

    /** The current subscription tier, or null if not subscribed. */
    val currentTier: SubscriptionTier?
        get() = subscriptionStatus.currentTier
}

/**
 * ViewModel managing billing state for the paywall and subscription screens.
 *
 * Coordinates between the [PurchaseService] and the UI, providing reactive
 * state updates for available plans, purchase progress, and subscription
 * status. Follows the same patterns as [SettingsViewModel] for state management.
 *
 * Usage:
 * ```kotlin
 * val purchaseViewModel: PurchaseViewModel = hiltViewModel()
 *
 * // In composable
 * val uiState by purchaseViewModel.uiState.collectAsState()
 * ```
 */
@HiltViewModel
class PurchaseViewModel @Inject constructor(
    private val purchaseService: PurchaseService
) : ViewModel() {

    private val _selectedBillingPeriod = MutableStateFlow(BillingPeriod.MONTHLY)
    private val _isLoading = MutableStateFlow(true)
    private val _isRestoring = MutableStateFlow(false)
    private val _showPurchaseSuccess = MutableStateFlow(false)
    private val _showRestoreSuccess = MutableStateFlow(false)
    private val _errorMessage = MutableStateFlow<String?>(null)

    /**
     * Combined UI state derived from purchase service flows and local state.
     */
    val uiState: StateFlow<PurchaseUiState> = combine(
        purchaseService.availablePlans,
        purchaseService.subscriptionStatus,
        purchaseService.isPurchasing,
        _selectedBillingPeriod,
        _isLoading
    ) { plans, status, isPurchasing, period, isLoading ->
        PurchaseUiState(
            availablePlans = plans,
            filteredPlans = plans.filter { it.billingPeriod == period },
            subscriptionStatus = status,
            selectedBillingPeriod = period,
            isPurchasing = isPurchasing,
            isLoading = isLoading && plans.isEmpty(),
            isRestoring = _isRestoring.value,
            showPurchaseSuccess = _showPurchaseSuccess.value,
            showRestoreSuccess = _showRestoreSuccess.value,
            errorMessage = _errorMessage.value
        )
    }.stateIn(
        scope = viewModelScope,
        started = SharingStarted.WhileSubscribed(5000),
        initialValue = PurchaseUiState()
    )

    init {
        connectAndLoadProducts()
    }

    // MARK: - Public Actions

    /**
     * Loads available products from Google Play.
     *
     * Connects to the billing service if needed, queries products,
     * and refreshes the current subscription status.
     */
    fun loadProducts() {
        connectAndLoadProducts()
    }

    /**
     * Purchases a subscription plan.
     *
     * Launches the Google Play purchase flow. The activity is required
     * to display the purchase dialog.
     *
     * @param activity The current activity for the purchase flow.
     * @param plan The subscription plan to purchase.
     */
    fun purchase(activity: Activity, plan: SubscriptionPlan) {
        viewModelScope.launch {
            _errorMessage.value = null

            when (val result = purchaseService.purchase(activity, plan)) {
                is PurchaseResult.Success -> {
                    _showPurchaseSuccess.value = true
                    purchaseService.refreshSubscriptionStatus()
                    Log.d(TAG, "Purchase successful for ${plan.id}")
                }

                is PurchaseResult.Cancelled -> {
                    Log.d(TAG, "Purchase cancelled by user")
                    // No error shown for user cancellation
                }

                is PurchaseResult.Pending -> {
                    _errorMessage.value = "Purchase is pending approval"
                    Log.d(TAG, "Purchase pending for ${plan.id}")
                }

                is PurchaseResult.Failed -> {
                    _errorMessage.value = result.message
                    Log.e(TAG, "Purchase failed: ${result.message}")
                }

                is PurchaseResult.BillingError -> {
                    _errorMessage.value = "Billing error: ${result.debugMessage}"
                    Log.e(TAG, "Billing error ${result.responseCode}: ${result.debugMessage}")
                }
            }
        }
    }

    /**
     * Restores previously purchased subscriptions.
     */
    fun restorePurchases() {
        viewModelScope.launch {
            _isRestoring.value = true
            _errorMessage.value = null

            try {
                val success = purchaseService.restorePurchases()
                if (success) {
                    val status = purchaseService.subscriptionStatus.value
                    if (status.isActive) {
                        _showRestoreSuccess.value = true
                    } else {
                        _errorMessage.value = "No active subscriptions found"
                    }
                } else {
                    _errorMessage.value = "Failed to restore purchases. Please try again."
                }
            } catch (e: Exception) {
                _errorMessage.value = "Error restoring purchases: ${e.localizedMessage}"
                Log.e(TAG, "Restore failed", e)
            } finally {
                _isRestoring.value = false
            }
        }
    }

    /**
     * Refreshes the current subscription status.
     */
    fun refreshStatus() {
        viewModelScope.launch {
            purchaseService.refreshSubscriptionStatus()
        }
    }

    /**
     * Changes the selected billing period filter.
     *
     * @param period The billing period to filter by.
     */
    fun selectBillingPeriod(period: BillingPeriod) {
        _selectedBillingPeriod.value = period
    }

    /**
     * Dismisses the purchase success dialog.
     */
    fun dismissPurchaseSuccess() {
        _showPurchaseSuccess.value = false
    }

    /**
     * Dismisses the restore success dialog.
     */
    fun dismissRestoreSuccess() {
        _showRestoreSuccess.value = false
    }

    /**
     * Dismisses the error message.
     */
    fun dismissError() {
        _errorMessage.value = null
    }

    // MARK: - Private

    /**
     * Connects to billing and loads products on initialization.
     */
    private fun connectAndLoadProducts() {
        viewModelScope.launch {
            _isLoading.value = true
            try {
                val connected = purchaseService.connect()
                if (connected) {
                    purchaseService.queryProducts()
                    purchaseService.refreshSubscriptionStatus()
                } else {
                    _errorMessage.value = "Unable to connect to Google Play"
                }
            } catch (e: Exception) {
                _errorMessage.value = "Error loading products: ${e.localizedMessage}"
                Log.e(TAG, "Failed to load products", e)
            } finally {
                _isLoading.value = false
            }
        }
    }

    override fun onCleared() {
        super.onCleared()
        // Note: Don't disconnect here as the service is a singleton
        // and may be used by other ViewModels
    }

    companion object {
        private const val TAG = "PurchaseViewModel"
    }
}

File: apps/mobile/ios/Happy/ViewModels/PurchaseViewModel.swift
//
//  PurchaseViewModel.swift
//  Happy
//
//  Copyright (c) 2024-2026 Enflame Media. All rights reserved.
//

import Foundation
import Combine

/// ViewModel for managing purchase state and coordinating between
/// the `PurchaseService` and purchase-related views.
///
/// Uses `ObservableObject` for iOS 16 compatibility.
final class PurchaseViewModel: ObservableObject {

    // MARK: - Published Properties

    /// Available subscription plans fetched from the App Store.
    @Published private(set) var plans: [SubscriptionPlan] = []

    /// The current subscription status.
    @Published private(set) var subscriptionStatus: SubscriptionStatus = .loading

    /// Whether products are currently being loaded.
    @Published private(set) var isLoading: Bool = false

    /// Whether a purchase is currently in progress.
    @Published private(set) var isPurchasing: Bool = false

    /// Whether a restore operation is in progress.
    @Published private(set) var isRestoring: Bool = false

    /// The currently selected billing period for the paywall.
    @Published var selectedBillingPeriod: BillingPeriod = .monthly

    /// Error message to display, if any.
    @Published var errorMessage: String?

    /// Whether to show the error alert.
    @Published var showError: Bool = false

    /// Whether to show the success alert after a purchase.
    @Published var showPurchaseSuccess: Bool = false

    /// Whether to show the restore success alert.
    @Published var showRestoreSuccess: Bool = false

    // MARK: - Computed Properties

    /// Plans filtered by the currently selected billing period.
    var filteredPlans: [SubscriptionPlan] {
        plans.filter { $0.billingPeriod == selectedBillingPeriod }
    }

    /// The recommended plan for the current billing period.
    var recommendedPlan: SubscriptionPlan? {
        filteredPlans.first { $0.isRecommended } ?? filteredPlans.first
    }

    /// Whether the user currently has an active subscription.
    var isSubscribed: Bool {
        subscriptionStatus.isActive
    }

    /// The current subscription tier, if subscribed.
    var currentTier: SubscriptionTier? {
        subscriptionStatus.currentTier
    }

    /// Whether to show the loading skeleton.
    var showLoadingSkeleton: Bool {
        isLoading && plans.isEmpty
    }

    // MARK: - Dependencies

    private let purchaseService: PurchaseProviding
    private var cancellables = Set<AnyCancellable>()

    // MARK: - Initialization

    /// Creates a new purchase view model.
    ///
    /// - Parameter purchaseService: The purchase service to use.
    ///   Defaults to the shared singleton.
    init(purchaseService: PurchaseProviding = PurchaseService.shared) {
        self.purchaseService = purchaseService
        setupSubscriptions()
    }

    // MARK: - Public Methods

    /// Loads available products from the App Store and refreshes subscription status.
    ///
    /// Call this when the paywall view appears.
    @MainActor
    func loadProducts() async {
        guard !isLoading else { return }
        isLoading = true
        errorMessage = nil

        do {
            try await purchaseService.loadProducts()
            plans = purchaseService.availablePlans
            await purchaseService.refreshSubscriptionStatus()
            subscriptionStatus = purchaseService.subscriptionStatus
        } catch let error as PurchaseError {
            // Don't show error for .notAvailable in simulator
            #if targetEnvironment(simulator)
            if error == .notAvailable {
                loadSamplePlans()
                isLoading = false
                return
            }
            #endif
            errorMessage = error.localizedDescription
            showError = true
        } catch {
            errorMessage = error.localizedDescription
            showError = true
        }

        isLoading = false
    }

    /// Purchases the specified subscription plan.
    ///
    /// - Parameter plan: The subscription plan to purchase.
    @MainActor
    func purchase(_ plan: SubscriptionPlan) async {
        guard !isPurchasing else { return }
        isPurchasing = true
        errorMessage = nil

        do {
            let success = try await purchaseService.purchase(plan)
            if success {
                subscriptionStatus = purchaseService.subscriptionStatus
                showPurchaseSuccess = true
            }
        } catch PurchaseError.purchaseCancelled {
            // User cancelled, no error to show
            #if DEBUG
            print("[PurchaseViewModel] User cancelled purchase")
            #endif
        } catch PurchaseError.pendingTransaction {
            errorMessage = PurchaseError.pendingTransaction.localizedDescription
            showError = true
        } catch {
            errorMessage = error.localizedDescription
            showError = true
        }

        isPurchasing = false
    }

    /// Restores previously purchased subscriptions.
    @MainActor
    func restorePurchases() async {
        guard !isRestoring else { return }
        isRestoring = true
        errorMessage = nil

        do {
            try await purchaseService.restorePurchases()
            subscriptionStatus = purchaseService.subscriptionStatus

            if subscriptionStatus.isActive {
                showRestoreSuccess = true
            } else {
                errorMessage = NSLocalizedString("subscription.restore.noPurchases", comment: "No purchases found to restore")
                showError = true
            }
        } catch {
            errorMessage = error.localizedDescription
            showError = true
        }

        isRestoring = false
    }

    /// Refreshes the current subscription status.
    @MainActor
    func refreshStatus() async {
        await purchaseService.refreshSubscriptionStatus()
        subscriptionStatus = purchaseService.subscriptionStatus
    }

    /// Dismisses the current error.
    @MainActor
    func dismissError() {
        errorMessage = nil
        showError = false
    }

    // MARK: - Private Methods

    /// Sets up Combine subscriptions to observe service state changes.
    private func setupSubscriptions() {
        // Observe the purchase service if it's a PurchaseService (ObservableObject)
        if let service = purchaseService as? PurchaseService {
            service.$subscriptionStatus
                .receive(on: DispatchQueue.main)
                .sink { [weak self] status in
                    self?.subscriptionStatus = status
                }
                .store(in: &cancellables)

            service.$availablePlans
                .receive(on: DispatchQueue.main)
                .sink { [weak self] plans in
                    self?.plans = plans
                }
                .store(in: &cancellables)

            service.$isPurchasing
                .receive(on: DispatchQueue.main)
                .sink { [weak self] purchasing in
                    self?.isPurchasing = purchasing
                }
                .store(in: &cancellables)
        }
    }

    /// Loads sample plans for use in simulator or preview mode.
    private func loadSamplePlans() {
        plans = [
            SubscriptionPlan.sampleProMonthly,
            SubscriptionPlan.sampleProAnnual
        ]
        subscriptionStatus = .notSubscribed
    }
}
